src/Chapter1.hs

```diff
@@ -553,8 +553,18 @@ value after "=" where the condition is true.
 
 Casual reminder about adding top-level type signatures for all functions :)
 -}
-
-mid x y z = error "mid: not implemented!"
+mid :: Ord a => a -> a -> a -> a
+mid x y z
+    | x > y && x < z = x
+    | x > z && x < y = x
+    | y > x && y < z = y
+    | y > z && y < x = y
+    | z > x && z < y = z
+    | z > y && z < x = z
+    | x == y && z /= x = x
+    | x == z && y /= x = x
+    | y == z && x /= y = y
+    | otherwise = x
```

Comment:
As we mentioned, the compiler in Haskell is very attentive to the exhaustive pattern-matching. And here it would warn you that `Pattern matching is not exhaustive`, as the guards have quite complicated logic, and the compiler won't be able to prove that it covers all the cases.

Because of that, you need to add another guard – `| otherwise = x`, to tell the compiler, that your pattern matching is exhaustive 🙂

However, you can use this last case for all guards that return `x`. As you for sure covered all other cases, so if the input won't satisfy any previous cases, you don't need to check conditions for `x`, it will be `x`.

```haskell
    | y > x && y < z = y
    | y > z && y < x = y
    | z > x && z < y = z
    | z > y && z < x = z
    | y == z && x /= y = y
    | otherwise = x
```